File: C/CommonLibraries/Common/Scripts/BatchScriptReports/Count.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Uheaa.Common.Scripts
{
	public class Count : IEquatable<Count>
	{
		
		public event ValueChanged NewValue;
		public delegate void ValueChanged(Count c);

		private int EojCount;

		public Count(int initalValue)
		{
			EojCount = initalValue;
		}

		public void Increment()
		{
			EojCount++;
			if (NewValue != null)
			{
				NewValue(this);
			}
		}

        public static Count operator ++(Count c)
        {
            c.Increment();
            return c;
        }

        public int Value
        {
            get
            {
                return EojCount;
            }
        }

        public static implicit operator int(Count c)
        {
            return (int)c.Value;
        }

		public override bool Equals(object obj)
		{
			Count other = (Count)obj;
			if (other == null)
			{
				return false;
			}
			else
			{
				return base.Equals(obj) && EojCount == other.EojCount;
			}
		}

		public override string ToString()
		{
			return EojCount.ToString();
		}

		#region IEquatable<T> Members

		public bool Equals(Count other)
		{
			return EojCount.Equals(other.EojCount);
		}

		public override int GetHashCode()
		{
			return EojCount.GetHashCode();
		}

		#endregion
	}
}

File: Q/Q/BatchScriptReports/Count.vb
Imports System

''' <summary>
''' Represents an integral value that can only be incremented.
''' </summary>
Public Class Count
    Implements IEquatable(Of Count)

    'Use an event to notify clients when the count changes.
    '(As of this class's creation date, this is needed by the EndOfJobReport class so it can write the new value to a backing store.)
    Public Event ValueChanged(ByVal newValue As Integer)

    Private _count As Integer

    Public Sub New(ByVal initialValue As Integer)
        _count = initialValue
    End Sub

    Public Sub Increment()
        _count += 1
        RaiseEvent ValueChanged(_count)
    End Sub

    'Allow integers to be compared to Count objects.
    Public Shared Operator =(ByVal a As Count, ByVal b As Integer) As Boolean
        If (CType(a, Object) Is Nothing) Then Return False
        Return a._count = b
    End Operator

    Public Shared Operator <>(ByVal a As Count, ByVal b As Integer) As Boolean
        Return (Not (a = b))
    End Operator

    Public Shared Operator >(ByVal a As Count, ByVal b As Integer) As Boolean
        If (CType(a, Object) Is Nothing) Then Return False
        Return a._count > b
    End Operator

    Public Shared Operator <(ByVal a As Count, ByVal b As Integer) As Boolean
        If (CType(a, Object) Is Nothing) Then Return False
        Return a._count < b
    End Operator

    'Override the standard stuff.
    Public Shared Operator =(ByVal a As Count, ByVal b As Count) As Boolean
        If (Object.ReferenceEquals(a, b)) Then Return True
        If (CType(a, Object) Is Nothing OrElse CType(b, Object) Is Nothing) Then Return False
        Return a._count = b._count
    End Operator

    Public Shared Operator <>(ByVal a As Count, ByVal b As Count) As Boolean
        Return (Not (a = b))
    End Operator

    Public Overrides Function Equals(ByVal obj As Object) As Boolean
        Dim other As Count = TryCast(obj, Count)
        If (other Is Nothing) Then Return False
        Return MyBase.Equals(obj) AndAlso Me._count = other._count
    End Function

    Public Overrides Function ToString() As String
        Return _count.ToString()
    End Function

#Region "IEquatable(Of Count) Members"

    Public Overloads Function Equals(ByVal other As Count) As Boolean Implements IEquatable(Of Count).Equals
        Return _count.Equals(other._count)
    End Function

    Public Overrides Function GetHashCode() As Integer
        Return _count.GetHashCode()
    End Function

#End Region
End Class
